lib/pos_staking_rewards.go

```diff
@@ -0,0 +1,116 @@
+	if err != nil {
+		return errors.Wrapf(err, "DistributeStakingRewardsToSnapshotStakes: problem retrieving snapshot stakes to reward: ")
+	}
+
+	// If there are no stakes to reward, then there's nothing to be done. Exit early here.
+	if len(snapshotStakesToReward) == 0 {
+		return nil
+	}
+
+	// Compute the total stake amount of all snapshot stakes, so we can determine the proportion of each
+	// staker's staked amount to the total.
+	snapshotStakesTotalStakeAmount := uint256.NewInt()
+	for _, snapshotStakeEntry := range snapshotStakesToReward {
+		snapshotStakesTotalStakeAmount.Add(snapshotStakesTotalStakeAmount, snapshotStakeEntry.StakeAmountNanos)
+	}
+
+	// Check if the sum of all of the stakes is zero. In practice this should never happen because it's not
+	// possible for a staker to stake zero DESO. We check it here to make this code more resilient, in case
+	// that assumption ever changes elsewhere in the codebase.
+	if snapshotStakesTotalStakeAmount.IsZero() {
+		return nil
+	}
+
+	// Loop through all of the snapshot stakes and reward them.
+	for _, snapshotStakeEntry := range snapshotStakesToReward {
+		rewardAmount := _computeStakingRewardAmount(
+			snapshotStakeEntry.StakeAmountNanos,
+			snapshotStakesTotalStakeAmount,
+			totalStakingRewards,
+		)
+		if rewardAmount.IsZero() {
+			continue
+		}
+
+		// At this point, we know that the staker has non-zero rewards. We need to determine how to
+		// distribute the rewards to them. We need to fetch their latest StakeEntry to determine
+		// whether they want to restake their rewards or not.
+
+		// Fetch the staker's latest StakeEntry.
+		stakeEntry, err := bav.GetStakeEntry(snapshotStakeEntry.ValidatorPKID, snapshotStakeEntry.StakerPKID)
+		if err != nil {
+			return errors.Wrapf(err, "DistributeStakingRewardsToSnapshotStakes: problem fetching staker's StakeEntry: ")
+		}
+
+		// At this point, there are three possible cases:
+		// 1. The staker still exists and wants to restake their rewards.
+		// 2. The staker still exists and does not want to restake their rewards.
+		// 3. The staker has unstaked since the snapshot was taken. They no longer have a
+		// StakeEntry. Their stake is currently in lockup.
+
+		// For case 1, we distribute the rewards by adding them to the staker's staked amount.
+		if stakeEntry != nil && stakeEntry.RestakeRewards {
+			stakeEntry.StakeAmountNanos.Add(stakeEntry.StakeAmountNanos, rewardAmount)
+			bav._setStakeEntryMappings(stakeEntry)
+			continue
+		}
+
+		// For cases 2 and 3, the staker no longer wants their rewards restaked. The staker is still
+		// eligible to receive rewards because the validator they had staked to was part of the validator
+		// set for the snapshot epoch. Their stake at the time was used to secure the network.
+
+		stakerPublicKey := bav.GetPublicKeyForPKID(snapshotStakeEntry.StakerPKID)
+
+		// TODO: _addBalance returns a UtxoOperation. Do we want to do anything with it?
```

Comment:
we should probably return an array of utxo operations from this function. similar to how we have a utxo operation when we delete the expired nonces, we'll want to save this state somehow.